### PR TITLE
React w2

### PR DESCRIPTION
**Pull request by elicarcar**
Project --> https://5de04ad5dd6ef9ce051bd80e--ecstatic-nobel-f62a26.netlify.com/
Exercises --> https://goofy-galileo-c4f7a7.netlify.com/

### CODE REVIEW

File: week2/react-exercises/src/components/Friend/index.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import Button from "../Button";
+import FriendProfile from "../FriendProfile";
+import uuid from "uuid/v1";
+import "./Friend.css";
+
+const Friend = () => {
+  const friendAPI = `https://www.randomuser.me/api?results=1`;
+  const [friend, setFriend] = useState({});
+  const [status, setStatus] = useState("loading");
+  const [errorMessage, setErrorMessage] = useState("");
```
Comment (Meno-101) [inline week2/react-exercises/src/components/Friend/index.js]: set status loading and set error message, same logic in multiple components, it's being said that it's same to say as the function `getFriend` it has:
 - fetch from an api.
 - check if the response is okay.
 - set data out of the response if no error.
 - set status with or without error.
 - set the error message if require.

These same exact steps has been accord in `Friend`,  `DogGallery` and `Joke` Components.
in each file they take 16 lines without the lines that specifies the request status and error message.

I see your way is the simplest way we could think of for the basic understanding of how react hooks works but I will add to that, that this way each time the component been re-rendered these functions will be recreated again and the useEffect will be re-invoked each time.
fortunately useEffect line 30 complains about what is happening in the browser's console, and that insures that it knows it has a function called `getSomething` as each of the previous mentioned.

So it's good for a learning progress, but if we can reduce the amount of code we have in the components and optimise a DRY way is gotta be a better readable code and better performance for what it's for a very small application a fractions of a millisecond which it's not important here.

if we broke this down into a small pieces we will find the following:
 - a status and an error message.
 - a function that fetch something from an Api.
 - And an always `useEffect` that uses the function.

as the list we have we can't move the `useEffect` outside the function unless we have something called `useReducer` which it's not part of this topic yet.

sticking with the list we can move point number one and two, as having them in the parent Component and that component will be our source to get most of the functionality we have in these children:
```jsx
import React, { useState } from "react"; // as what I think.
import "./App.css";
import Friend from "./components/Friend";
import DogGallery from "./components/DogGallery";
import RandomJoke from "./components/RandomJoke";

function App() {
  const [status, setStatus] = useState('Loading');
  const [errorMessage, setErrorMessage] = useState('');

  // the fetch functionality along with the url and setters as `setFriend` using parameters.

  return ( // now each of these may have something similar to the first.
    <div className="App">
      <Friend errorMessage={errorMessage} status={status} fetchUrl={/* your fetch function 🙂*/} />
      <DogGallery />
      <RandomJoke />
    </div>
  );
}

export default App;
```

After all of that I'm sure you will agree with me that this will be a lot cleaner.
Please have a look, try it out if you find it very complicated don't waste your time just know that React been invented for DRY code and that can do a lot more than what we was doing in the past using vanilla javascript 🙂

File: week2/react-exercises/src/components/Friend/index.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        setErrorMessage(error.message);
+      }
+    }
+  }
+
+  useEffect(() => {
+    getFriend();
+    return () => {
+      getFriend();
+    };
```
Comment (Meno-101) [inline week2/react-exercises/src/components/Friend/index.js]: as what is been specified in the documentation we don't need this cleaning up after using the `useEffect` hook in this case.
> If you’re familiar with React class lifecycle methods, you can think of useEffect Hook as `componentDidMount`, `componentDidUpdate`, and `componentWillUnmount` combined.

the cleanup part is only if you are expecting the component to go away while the program is running as a remove button that will remove an element, this element should be cleaned up if it has anything registered as `setTimeout`, `setInterval` service worker registrations etc...

Making this right we can look to line 29 of `RandomJoke` component.

File: week2/react-exercises/src/components/RandomJoke/index.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  }
+
+  useEffect(() => {
+    setInterval(async () => getJoke(), 7000);
+
+    return () => {
+      console.log("And its over");
+      return getJoke();
+    };
+  }, []);
```
Comment (Meno-101) [inline week2/react-exercises/src/components/RandomJoke/index.js]: This has to be cleaned, the difference is simple, cleaning up is the way to remove any registered memory space out of the application as `removeEventListener` or maybe as `clearInterval`, `clearTimeout`.

and that happen exactly right before the element been omitted out of the `DOM`, in other words we don't need to make a fetch in line 34 before the `DOM` removes this element, just it doesn't make sense to remove something you will not display on anymore and fetch for it.

in short, what have to be done is this:
```jsx
  useEffect(() => {
    const fn = async () => getJoke(); // we are making sure that this function will be referenced to be cleared.
    setInterval(fn, 7000); // setting the interval

    return () => {
      console.log("And its over"); // clearing when it's needed.
      clearInterval(fn);
    };
  }, [getJoke]);
```
